tui/network.py

```diff
@@ -19,128 +19,326 @@
-    vlan_cb.setCallback(use_vlan_cb_change, ())
-    if defaults and defaults.isVlan():
-        vlan_field.set(str(defaults.vlan))
-    else:
-        vlan_field.setFlags(FLAG_DISABLED, False)
-
-    ip_text = Textbox(15, 1, "IP Address:")
-    subnet_text = Textbox(15, 1, "Subnet mask:")
-    gateway_text = Textbox(15, 1, "Gateway:")
-    dns_text = Textbox(15, 1, "Nameserver:")
-    vlan_text = Textbox(15, 1, "VLAN (1-4094):")
-
-    entry_grid = Grid(2, include_dns and 4 or 3)
-    entry_grid.setField(ip_text, 0, 0)
-    entry_grid.setField(ip_field, 1, 0)
-    entry_grid.setField(subnet_text, 0, 1)
-    entry_grid.setField(subnet_field, 1, 1)
-    entry_grid.setField(gateway_text, 0, 2)
-    entry_grid.setField(gateway_field, 1, 2)
-    if include_dns:
-        entry_grid.setField(dns_text, 0, 3)
-        entry_grid.setField(dns_field, 1, 3)
-
-    vlan_grid =  Grid(2, 1)
-    vlan_grid.setField(vlan_text, 0, 0)
-    vlan_grid.setField(vlan_field, 1, 0)
-
-    gf.add(text, 0, 0, padding=(0, 0, 0, 1))
-    gf.add(dhcp_rb, 0, 2, anchorLeft=True)
-    gf.add(static_rb, 0, 3, anchorLeft=True)
-    gf.add(entry_grid, 0, 4, padding=(0, 0, 0, 1))
-    gf.add(vlan_cb, 0, 5, anchorLeft=True)
-    gf.add(vlan_grid, 0, 6, padding=(0, 0, 0, 1))
-    gf.add(buttons, 0, 7, growx=1)
-
-    loop = True
-    while loop:
-        result = gf.run()
-
-        if buttons.buttonPressed(result) in ['ok', None]:
-            # validate input
-            msg = ''
-            if static_rb.selected():
-                if not netutil.valid_ip_addr(ip_field.value()):
-                    msg = 'IP Address'
-                elif not netutil.valid_ip_addr(subnet_field.value()):
-                    msg = 'Subnet mask'
-                elif gateway_field.value() != '' and not netutil.valid_ip_addr(gateway_field.value()):
-                    msg = 'Gateway'
-                elif dns_field.value() != '' and not netutil.valid_ip_addr(dns_field.value()):
-                    msg = 'Nameserver'
-            if vlan_cb.selected():
-                if not netutil.valid_vlan(vlan_field.value()):
-                    msg = 'VLAN'
-            if msg != '':
-                tui.progress.OKDialog("Networking", "Invalid %s, please check the field and try again." % msg)
+        tui.screen.popWindow()
+
+        if buttons.buttonPressed(result) == 'back': return LEFT_BACKWARDS, None
+
+        vlan_value = int(vlan_field.value()) if vlan_cb.selected() else None
+        if bool(dhcp_rb.selected()):
+            answers = NetInterface(NetInterface.DHCP, nic.hwaddr, vlan=vlan_value)
+        else:
+            answers = NetInterface(NetInterface.Static, nic.hwaddr, ip_field.value(),
+                                subnet_field.value(), gateway_field.value(),
+                                dns_field.value(), vlan=vlan_value)
+        return RIGHT_FORWARDS, answers
+
+    def get_ipv6_configuration(nic, txt, defaults, include_dns):
```

Comment:
far too long and far too complicated

Reply:
Not sure how to do better than a subfunction per screen.
TUI in code is very verbose.

Reply:
I'v mutualized the codes for both IPv4 and IPv6 interface configuration form